File: java/src/main/java/com/leetcode/twosum/TwoPassHashTableSolution.java
package com.leetcode.twosum;

import java.util.HashMap;
import java.util.Map;

public class TwoPassHashTableSolution implements TwoSumSolver {
  public int[] twoSum(int[] nums, int target) {
    Map<Integer, Integer> map = new HashMap<>();

    // First pass: build the hash map
    for (int i = 0; i < nums.length; i++) {
      map.put(nums[i], i);
    }

    // Second pass: find the complement
    for (int i = 0; i < nums.length; i++) {
      int complement = target - nums[i];
      if (map.containsKey(complement) && map.get(complement) != i) {
        return new int[] { i, map.get(complement) };
      }
    }

    // If no valid pair is found, return an empty array
    return new int[] {};
  }
}

File: typescript/src/twosum/twopass-hashtable-solution.ts
import { TwoSumSolver } from "./types";

export const twoPassHashTableSolution: TwoSumSolver = (nums, target) => {
  const map: Map<number, number> = new Map();
    for (let i = 0; i < nums.length; i++) {
        map.set(nums[i], i);
    }
    for (let i = 0; i < nums.length; i++) {
        const complement = target - nums[i];
        const complementindex = map.get(complement);
        if (complementindex !== undefined && complementindex !== i) {
            return [i, complementindex];
        }
    }
    // If no valid pair is found, return an empty array
    return [];
};